lib/pbench/test/functional/server/test_put.py

```diff
@@ -106,3 +130,24 @@ def test_index_all(self, server_client: PbenchServerClient, login_user):
         assert count == len(
             self.tarballs
         ), f"Didn't find all expected datasets, found {count} of {len(self.tarballs)}"
+
+    def test_delete(self, server_client: PbenchServerClient, login_user):
+        """Verify we can delete each previously uploaded dataset."""
+        datasets = server_client.get_list()
+        datasets_hash = {}
+        try:
+            for dataset in datasets:
+                datasets_hash[f"{dataset.name}.tar.xz"] = dataset.resource_id
+        except HTTPError as exc:
```

Comment:
I don't see anything in that `try` block which could raise `HTTPError`...am I missing it?

Did you want this `try` block around the call to `server_client.get_list()`?  (I wouldn't bother, since the test will fail just fine without it....)

[Also, you really do love f-strings, don't you!?  😆  We didn't really need one at line 120, and we don't really need one here -- at line 120, the `print()` function will do a fine job of joining the stings, and concatenation will work quite nicely here.  🙂]

Reply:
Generators...

Reply:
Once burned, twice shy...perhaps, now that I've been burned _twice_ I'll be shy enough to remember this.

I think the problem is having a function return a generator which is then assigned to a local variable -- that creates a double pitfall:  first, the local variable doesn't look like a generator, so the reader is quick to assume that it's a list or other iterable collection; then, later, the implementor (or someone else modifying the code) tries to reuse the generator without fetching a new one and wonders why the iteration doesn't produce any results (because they were all used up in the previous loop).

So, I think it's fine to create and use generators, but I think it would be better to reference the generator-producing function directly in the `for` loop rather than caching the generator in a local variable -- that would cover both pitfalls.